### PR TITLE
ENH: Finish implementing generator-based incremental multi-tau correlation

### PR DESCRIPTION
**Pull request by ericdill**
supersedes #368 
example notebook: scikit-beam/scikit-beam-examples#43

### CODE REVIEW

**Comment by ericdill** [comment]
attn @tacaswell 

This PR crept in scope so that I just ended up finishing the implementation of the partial multi-tau generator based implementation.  I am punting on cythonizing the multi-tau implementation until this PR is in.

Companion PR is scikit-beam/scikit-beam-examples#43

File: skxray/core/accumulators/correlation/correlation.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        label_mask,
+        track_level,
+        cur,
+        pixel_list,
+        label_mapping,
+    )
+
+
+def lazy_multi_tau(image_iterable, num_levels, num_bufs, labels,
+                   processing_func=None, _state=None):
```
Comment (tacaswell) [inline skxray/core/accumulators/correlation/correlation.py]: I do not remember if I initially suggested it, but I am not a fan of the `_` arg.

File: skxray/core/accumulators/correlation/correlation.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        # If any past intensities are zero, then g2 cannot be normalized at
+        # those levels. This if/else code block is basically preventing
+        # divide-by-zero errors.
+        if len(np.where(s.past_intensity == 0)[0]) != 0:
+            g_max = np.where(s.past_intensity == 0)[0][0]
+        else:
+            g_max = s.past_intensity.shape[0]
+
+        # Normalize g2 by the product of past_intensity and future_intensity
+        g2 = (s.G[:g_max] /
```
Comment (tacaswell) [inline skxray/core/accumulators/correlation/correlation.py]: Is there a way to defer this computation unless it is really needed?

File: skxray/core/accumulators/correlation/pyprocess.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)   #
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,  #
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OTHERWISE) ARISING   #
+# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE   #
+# POSSIBILITY OF SUCH DAMAGE.                                          #
+from __future__ import absolute_import, division, print_function
+import numpy as np
+
+
+def pyprocess(buf, G, past_intensity_norm, future_intensity_norm,
```
Comment (tacaswell) [inline skxray/core/accumulators/correlation/pyprocess.py]: This needs a better name.  I do not think this needs to be in it's own file.  I see the symmetry argument against the (eventual) stand-alone cython file, but eh.

Comment (ericdill) [inline skxray/core/accumulators/correlation/pyprocess.py]: This is now called `_one_time_process`

File: skxray/core/accumulators/correlation/correlation.py (hunk 2 above)
Comment (ericdill) [inline skxray/core/accumulators/correlation/correlation.py]: There is little gain in this.  As per this question, I benchmarked the time savings for computing versus not computing g2 here.  The notebook can be found [here] (https://gist.github.com/ericdill/f6b54927b184b6a47be9) and the results show that there is about a ~2% speed up in not computing g2 for very small ROIs and no noticeable difference if the ROIs are very large.

File: skxray/core/accumulators/correlation/correlation.py (hunk 1 above)
Comment (ericdill) [inline skxray/core/accumulators/correlation/correlation.py]: Renamed `_state` to `internal_state`

**Comment by tacaswell** [comment]
:+1: from me

@sameera2004 Can you merge this?

**Comment by tacaswell** [comment]
@sameera2004 predicated on if you agree it is correct.

**Comment by sameera2004** [comment]
I am trying to go through your notebook example and try to run with different data set. However, there is a merge conflict in this PR, which don't allow me to merge

**Comment by ericdill** [comment]
@sameera2004 I am confused where you are getting a merge conflict.  I see no merge conflicts in #386 or scikit-beam/scikit-beam-examples#43.

**Comment by ericdill** [comment]
How did you check out and install the branch on scikit-beam and the branch on scikit-beam-examples?

**Comment by sameera2004** [comment]
@tacaswell fixed the issue, now I will check it, both scikit-beam and scikit-beam-examples are fine now, somehow you turned on a flag not allowing to merge

**Comment by ericdill** [comment]
Ah ok, sorry about that!  Thanks for looking this over

File: skxray/core/correlation.py [hunk 4]
```diff
@@ -108,243 +250,104 @@ def multi_tau_auto_corr(num_levels, num_bufs, labels, images):
...
-                prev = 1 + (cur[level - 1] - 2) % num_bufs
-                cur[level] = 1 + cur[level] % num_bufs
+                prev = (1 + (s.cur[level - 1] - 2) % num_bufs)
+                s.cur[level] = (
+                    1 + s.cur[level] % num_bufs)
 
-                buf[level, cur[level] - 1] = (buf[level - 1, prev - 1] +
-                                              buf[level - 1,
-                                                  cur[level - 1] - 1])/2
+                # TODO clean this up. it is hard to understand
```
Comment (sameera2004) [inline skxray/core/correlation.py]: @ericdill here the frames add average according to the multi tau levels and buffers

**Comment by sameera2004** [comment]
Looks good

**Comment by ericdill** [comment]
Awesome, thanks @sameera2004 and @tacaswell !

**Pull request merged**